wiring/src/spark_wiring_cloud_event.cpp

```diff
@@ -0,0 +1,778 @@
+    if (!isWritable()) {
+        int err = error();
+        if (err < 0) {
+            return err;
+        }
+        return Error::INVALID_STATE;
+    }
+    auto payload = getValidPayload();
+    if (!payload) {
+        return error();
+    }
+    int r = coap_write_payload(payload, data, size, d_->pos, nullptr /* reserved */);
+    if (r < 0) {
+        if (r == Error::COAP_TOO_LARGE_PAYLOAD) {
+            LOG(ERROR, "Event data is too large");
+            return setFailed(r);
+        }
+        LOG(ERROR, "coap_write_payload() failed: %d", r);
+        return setInvalid(r);
+    }
+    d_->pos += r;
+    return r;
+}
+
+int CloudEvent::size(size_t size) {
+    if (!isWritable()) {
+        int err = error();
+        if (err < 0) {
+            return err;
+        }
+        return Error::INVALID_STATE;
+    }
+    auto payload = getValidPayload();
+    if (!payload) {
+        return error();
+    }
+    int r = coap_set_payload_size(payload, size, nullptr /* reserved */);
+    if (r < 0) {
+        if (r == Error::COAP_TOO_LARGE_PAYLOAD) {
+            LOG(ERROR, "Event data is too large");
+            return setFailed(r);
+        }
+        LOG(ERROR, "coap_set_payload_size() failed: %d", r);
+        return setInvalid(r);
+    }
+    if (d_->pos > size) {
+        d_->pos = size;
+    }
+    return 0;
+}
+
+size_t CloudEvent::size() const {
+    if (!isReadable() || !d_->payload) {
+        return 0;
+    }
+    int r = coap_get_payload_size(d_->payload.get(), nullptr /* reserved */);
+    if (r < 0) {
+        LOG(ERROR, "coap_get_payload_size() failed: %d", r);
+        return 0;
+    }
+    return r;
+}
+
+int CloudEvent::pos(size_t pos) {
+    // The current position is used both when reading and writing so using the least restrictive
+    // status check here
+    if (!isReadable()) {
+        return error();
+    }
+    size_t size = 0;
+    if (d_->payload) {
+        int r = coap_get_payload_size(d_->payload.get(), nullptr /* reserved */);
+        if (r < 0) {
+            LOG(ERROR, "coap_get_payload_size() failed: %d", r);
+            return r;
+        }
+        size = r;
+    }
+    if (pos > size) {
+        pos = size;
+    }
+    d_->pos = pos;
+    return pos;
+}
+
+size_t CloudEvent::pos() const {
+    if (!isReadable()) {
+        return 0;
+    }
+    return d_->pos;
+}
+
+CloudEvent& CloudEvent::operator=(CloudEvent event) {
+    swap(*this, event);
+    return *this;
+}
+
+int CloudEvent::publish() {
+    auto status = this->status();
+    if (status == Status::INVALID) {
+        return error();
+    }
+    if (status == Status::SENDING) {
+        LOG(ERROR, "Event is being sent already");
+        return Error::INVALID_STATE;
+    }
+    if (!d_->name) {
+        LOG(ERROR, "Event name is missing");
+        return setFailed(Error::INVALID_STATE);
+    }
+    size_t size = this->size();
+    if (!RateLimiter::instance().take(size)) {
+        return Error::LIMIT_EXCEEDED;
```

Comment:
So exceeding the amount of data in flight will return LIMIT_EXCEEDED, but the event will stay in the NEW state so it can be sent again later. Is there a way to query how much data is in flight to know when a new event can be sent?

Reply:
> the event will stay in the NEW state

That was an oversight. The event will transition to a failed state on this error as well.

> Is there a way to query how much data is in flight to know when a new event can be sent?

There's `canPublish()`:
```
if (CloudEvent::canPublish(3000)) {
    Particle.publish(CloudEvent().name("my_event").data(buf, 3000));
}
```

I was hesitant to define it as `Particle.canPublish()` as not to create confusion since we have different limits for the new and old APIs.